src/www/Inset.ts

```diff
@@ -0,0 +1,266 @@
+/*
+   Copyright 2022 Total Pave Inc.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+export interface IInsetAPI {
+    addListener: (callback: IInsetCallbackFunc) => void;
+    removeListener: (callback: IInsetCallbackFunc) => void;
+    getInsets: () => IInset;
+}
+
+declare global {
+    interface ITotalpave {
+        Insets: IInsetAPI;
+    }
+    interface Window {
+        totalpave: ITotalpave;
+    }
+}
+
+export const SERVICE_NAME: string = "Insets";
+
+export interface IInset {
+    top: number;
+    right: number;
+    bottom: number;
+    left: number;
+}
+
+interface IInsetEvent<T = unknown> {
+    type: 'init' | 'update';
+    data: T;
+}
+
+type IInsetInitEvent = IInsetEvent<string>;
+interface IInsetUpdateEvent extends IInsetEvent<IInset> {
+    id: string;
+}
```

Comment:
Move id to the base IInsetEvent.
I don't see why any event would be missing the id attribute, including the init event.
data should be a nullable attribute (null for init). Data should remain as a generic type.

init event will not use data and will instead use the id attribute.
IInsetUpdateEvent type the data to IInset and use data.

Reply:
the id is passed back as `data` in the init event. The entire purpose of this event is send back an native identifier back to the webview so it can finish constructing the `Inset` instance.

the `id` field on the update event is there because at that time it's expected that the `Inset` object is already established and can be retrieved by the id. Whereas during the init event it's in the process of constructing that inset instance.

So I don't really agree with this suggestion.